addons/delivery/models/delivery_carrier.py

```diff
@@ -154,6 +178,22 @@ def _match_address(self, partner):
                 return False
         return True
 
+    def _match_must_have_tags(self, order):
+        self.ensure_one()
+        return all(tag in order.order_line.product_id.all_product_tag_ids for tag in self.must_have_tag_ids)
+
+    def _match_excluded_tags(self, order):
+        self.ensure_one()
+        return not any(tag in order.order_line.product_id.all_product_tag_ids for tag in self.excluded_tag_ids)
+
+    def _match_weight(self, order):
+        self.ensure_one()
+        return not self.max_weight or sum(order_line.product_id.weight * order_line.product_uom_qty for order_line in order.order_line) <= self.max_weight
```

Comment:
Product weight represent 1 "unit" of the product default uom, but the quantity on the sale.order.line may be another one. Same for the volume